File: 250713/0300/minkyu.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);
        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }

        return *max_element(dp.begin(), dp.end());
    }

    int lengthOfLIS2(vector<int>& nums) {
        vector<int> lis { nums[0] };
        for (int num : nums) {
            auto it = lower_bound(lis.begin(), lis.end(), num);
            if (it == lis.end()) lis.push_back(num);
            else *it = num;
        }
        return lis.size();
    }
};

File: 250713/0053/minkyu.py
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        max_sum = nums[0]
        temp_sum = nums[0]
        for num in nums[1:]:
            temp_sum = max(temp_sum + num, num)
            if temp_sum > max_sum:
                max_sum = temp_sum
        return max_sum